### ISSUE TITLE
cabal init logs at loglevel "Log"?

### ISSUE THREAD

**Issue by emilypi** [haskell/cabal#9151]
On 3.10, but appears to be introduced by @ptkato in 3.8: 

```
Add informative comments to each field in the cabal file. (y/n)? [default: y] n
[Log] Using cabal specification: 3.4
[Log] Creating fresh file LICENSE...
[Log] Creating fresh file CHANGELOG.md...
[Log] Creating fresh directory ./src...
[Log] Creating fresh file src/MyLib.hs...
[Log] Creating fresh directory ./app...
[Log] Creating fresh file app/Main.hs...
[Log] Creating fresh directory ./test...
[Log] Creating fresh file test/Main.hs...
[Log] Creating fresh file arithmetic.cabal...
[Warning] No synopsis given. You should edit the .cabal file and add one.
[Info] You may want to edit the .cabal file and add a Description field.
```

This is a quick change to `FileCreators.hs` and removing the ADT entry for "Log" in the loglevel type. This is a good first issue.

**Comment by aleeusgr** [comment]
Hi.
Is anyone available to elaborate?
1. I search for the mentioned file, I find not what I expected:
```shell
✔ ~/workshop/cabal [master|✔] 
08:00 $ find . -name "FileCreators.hs"
./cabal-install/tests/UnitTests/Distribution/Client/Init/FileCreators.hs
./cabal-install/src/Distribution/Client/Init/FileCreators.hs
```
2. What is a testable expectation? How do I know I made the right change?
I should test with `cabal init` and my expectation is to see less log lines?

Thank you!

**Comment by emilypi** [comment]
we can work on it together if you like! This is a fairly easy one with a fairly opaque test strategy :)

**Comment by aleeusgr** [comment]
I'd like that, thanks. 
I cloned the repo and skimmed the contributing guide.
I wanted to open PR into my fork to share around and later open PR to the main repo.

I will try to understand what `FileCreators.hs` refers to and report back if I don't have success.

**Comment by aleeusgr** [comment]
@emilypi plz tell me more about the testing strategy. What do you mean?
🙇‍♂️

**Comment by emilypi** [comment]
ah you're referring to the cabal test suite not the `cabal-install` test suite. We'd just need to update `./cabal-install/tests/UnitTests/Distribution/Client/Init/FileCreators.hs` and test the outputs. I don't believe we test the loglevel, so we'll have to decide whether we care to, or leave it as an in-place fix without tests

**Comment by aleeusgr** [comment]
- [x] find `cabal-install` 3.10 test suite. (?)
  - [x] search `repo:haskell/cabal [Log]`
  - [x] search `repo:haskell/cabal user:ptkato`
  - [x] https://github.com/haskell/cabal/tree/master/release-notes
  - [x] test cabal-install

<details><summary>Details</summary>
<p>

> Error: cabal: Tests failed for test:long-tests from cabal-install-3.11.0.0.
3 out of 90 tests failed (24.30s)
Test suite long-tests: FAIL
Test suite logged to:

</p>
</details> 

🤔 I just noticed the OP looks like its about `cabal init`, why am I looking at `cabal-install`?
but anyway, isn't there supposed to be a logging module that is reused?
How are logs created?
> Test suites using the `exitcode-stdio-1.0` interface ([src](https://cabal.readthedocs.io/en/3.4/cabal-package.html?highlight=log#test-suites))

```shell
08:10 $ uptime
08:12:31  up   0:46,  2 users,  load average: 0.31, 0.40, 0.43
```

**Comment by emilypi** [comment]
@Mikolaj do you think this is a bot? I'm thoroughly confused by what's going on here.

**Comment by aleeusgr** [comment]
> @Mikolaj do you think this is a bot? I'm thoroughly confused by what's going on here.

I'm sorry for confusing you. Being able to find things quickly is something I am practicing, I am confused as well.
I looked into the file you suggested but could not make understand what's going on. 
I tired using github search, but it didn't work. 

In one of the files I found something that looks not far from what I am looking for, and then I got distracted ( I am an enthusiast and I don't have full day to attend to this kind of work), so I continue tomorrow,

Thanks for caring🙇‍♂️

**Comment by emilypi** [comment]
Everything you need is in this directory: https://github.com/haskell/cabal/tree/master/cabal-install/src/Distribution/Client/Init

Try `Types.hs`: the logger for `init` is implemented there.

**Comment by ptkato** [comment]
for context, the motivation behind the "log" was just to keep up with the pattern of having a lil' tag for each type of output within init

**Comment by aleeusgr** [comment]
> Everything you need is in this directory: https://github.com/haskell/cabal/tree/master/cabal-install/src/Distribution/Client/Init
> 
> Try `Types.hs`: the logger for `init` is implemented there. Feel free to ask questions instead of rushing off and spamming the issue with nonsensical analysis 😅. We're sociable people here at Cabal.

Thank you for being so welcoming, I feel like I want to stay already.

**Comment by aleeusgr** [comment]
Did I find the right line?

https://github.com/haskell/cabal/blob/b0c52fc906352e33e5a50a4ef995dbd685f40989/cabal-install/src/Distribution/Client/Init/Types.hs#L371

Thx.

**Comment by emilypi** [comment]
Yeah it'll be there and the actual ADT for severity should probalby change because "Log" is useless as a datatype, and the show instance should probably be made explicit (or better yet, erase the show instance and write a function `displaySeverity` or whatever and shim that in instead of `show`).

**Comment by aleeusgr** [comment]
What `displaySeverity` should do?

I thought at first that the problem was the redundant [Log] in front of every message and the goal is just to remove that.
Sorry for being dumb, but I'd rather figure out what is needed before go about trying to build something.

**Comment by emilypi** [comment]
No, the problem is that "Log" is not a [loglevel](https://betterstack.com/community/guides/logging/log-levels-explained/). It should probably display `Info`, but for some reason `Log` was chosen. Arguably it shouldn't even be an entry in the ADT and the severity should be at `Info` for everything that's not a `Warn`. Personally, i think it would be best to now confuse the `Show` instance with the textual representation of the loglevel, since it's a bit of an abuse of `Show`, for which best practice is usually to only use it for debugging purposes. That's why i'm suggesting a separate function that displays the severity.

**Comment by emilypi** [comment]
So concretely, we need to: 

1. Remove the loglevel entry for `Log`, replacing places where it was used with `Info` (presumably, to be reviewed later)
2. Not defer to `Show`, deleting the `Show` instance for `Severity`.
3. Write a textual display function for the loglevels in the place of the uses of `show`.

**Comment by aleeusgr** [comment]
Ok, this kinda makes sense, but also kinda does not: I need to understand Haskell and the code base a bit better.

Video is great, thanks a lot! I don't want to waste your time teaching me things I can learn myself:  I'll dig around docs and code async, if you don't mind.

But see you soon, hopefully.

**Comment by emilypi** [comment]
Believe it or not, I *like* mentoring people 😂

**Comment by aleeusgr** [comment]
I know what you mean: I am an instructor in kiteboarding, and (maybe next week) a teacher of maths and CS for kids in a homeschooling community, 

Learning on my own seems less efficient, but I enjoy being confused for a while: somehow it makes any accomplishment more rewarding. 
I try not to overdo.

**Comment by aleeusgr** [comment]
Today I was distracted more than yesterday, but I think I got what you meant in 1:
https://github.com/aleeusgr/cabal/pull/2
Please confirm that I did what you expected. 🙇‍♂️
Thx

I hoped to do more but I got a student today and now I am 🤯
I will look into 2. first thing tomorrow morning.

**Comment by emilypi** [comment]
Yep that takes care of that :)

**Comment by aleeusgr** [comment]
It seems like 2 and 3 are not separable: deleting `Show` from the ADT's deriving causes errors where `show` is used, so the next step is to replace it with `displaySeverity`

`displaySeverity` should be just a wrapper for `show` and its main purpose is to give a more descriptive name to the action, right? 

I should not delete `Log` from `Severity` in Types , right?

**Comment by ptkato** [comment]
`Log` serves no purpose, I don't recall why I put it there, but it should be removed. About `displaySeverity`, nah, `show` shouldn't be involved with it,  as mentioned, `Show` is mostly intended and used for debugging stuff, for actual displaying we want, well, an actual displaying function.

By the by, the square brackets in
https://github.com/haskell/cabal/blob/b0c52fc906352e33e5a50a4ef995dbd685f40989/cabal-install/src/Distribution/Client/Init/Types.hs#L371
could also be moved into the `displaySeverity` function, but I guess that's more a matter of preference.

**Comment by aleeusgr** [comment]
Removed `Log` from Severity in Types.hs. It was used in Utils.hs. I replaced it with `Info`

https://github.com/aleeusgr/cabal/pull/2/files

I see usage of `prettyShow` across the repo. 
Should I avoid using it as well? 
Is it ok to just replace `show` with `prettyShow`?  
Should I replace `show` with `displaySeverity` and have `prettyShow` under the hood?

thx

**Comment by ptkato** [comment]
You could use `prettyShow`, maybe, but then you'd need to make an instance of `Pretty` for `Severity`.
https://github.com/haskell/cabal/blob/9f373253b9032ae32ead9f5ca17e49d192f9b8b9/Cabal-syntax/src/Distribution/Pretty.hs#L43

**Comment by aleeusgr** [comment]
@emilypi I'd like to look into this together, if you don't mind. 
I think I got myself confused enough)

I don't understand how to avoid using Show: I look into Learn you a Haskell and it is mentioned as third most used typeclass.
As we discussed above I could use prettyShow but making an instance of Pretty for Severity isn't something I understand how to do: I tried modifying Pretty.hs or adding an instance in `Init/Types.hs` but in both cases I don't get what is the next step to take me closer to the goal.

Beside that `show severity` is used twice in Types, L371 and L410 - symbols added to the string are different in each case, maybe `displaySeverity` should not add brackets by default, as suggested above? 

Thank you.

**Comment by aleeusgr** [comment]
Hi! This is getting delayed! I created a small [PR.](https://github.com/haskell/cabal/pull/9346)

I am up for a call if you'd like.

I don't understand the idea about `Show`, its used all over the place.

 I seek career advice: I'm stuck. Too old for Junior roles, too inexperienced for Senior, and I get turned down without comments - I look for low hanging fruit I could improve to increase my chances of getting a job.

Thx

**Comment by emilypi** [comment]
Sorry i've been a little tied up - this is basic haskell - let's just review on the PR so we can talk with examples.

**Comment by aleeusgr** [comment]
NP. I opened PR because it's contents fit the title of the issue.

Thanks for your patience and for mentoring me on basic Haskell: I learn QuickCheck and how to model stuff but I've come to Hacktoberfest to get exposure with Haskell in general. 

I got stuck: `Show` is used quite a few times and I don't get what to replace it with.

I understand `displaySeverity` should produce a string.

It seems it's input type should be `Severity`.

With this making an instance of `Pretty` for severity looked as a good option but it goes beyond the scope of what we talked about in the beginning.

Thank you for your comments 🙏

**Comment by aleeusgr** [comment]
NP. I opened PR because it's contents fit the title of the issue.

Thanks for your patience and for mentoring me on basic Haskell: I learn QuickCheck and how to model stuff but I've come to Hacktoberfest to get exposure with Haskell in general. 

I got stuck: `Show` is used quite a few times and I don't get what to replace it with.

I understand `displaySeverity` should produce a string.

It seems it's input type should be `Severity`.

With this making an instance of `Pretty` for severity looked as a good option but it goes beyond the scope of what we talked about in the beginning.

Thank you for your comments 🙏
